pkg/daemon/daemon.go

```diff
@@ -195,15 +195,31 @@ func (dn *Daemon) isDesiredMachineState() (bool, error) {
 		return false, err
 	}
 
+	isDesiredOS, err := dn.checkOS(desiredConfig.Spec.OSImageURL)
+	if err != nil {
+		return false, err
+	}
+
 	if dn.checkFiles(desiredConfig.Spec.Config.Storage.Files) &&
-		dn.checkUnits(desiredConfig.Spec.Config.Systemd.Units) {
+		dn.checkUnits(desiredConfig.Spec.Config.Systemd.Units) &&
+		isDesiredOS {
```

Comment:
nit: My brain does not like how this `if` statement is structured, but it's not problematic.